pandas/core/algorithms.py

```diff
@@ -362,7 +363,14 @@ def unique(values):
     htable, _, values, dtype, ndtype = _get_hashtable_algo(values)
 
     table = htable(len(values))
-    uniques = table.unique(values)
+
+    if isinstance(values, ABCSparseArray):
```

Comment:
not sure you saw my comment. this should not be handling unique like this. SparseArray should have a method ``.unique()`` which  can directly handle the unique (and then call algorithms.unique on the sparse values).

Reply:
IOW you might be able to get aways with doing a 
```
if hasattr(values, 'unique'):
    return values.unique()
...
```
need to avoid recursion, but here values must be a ndarray or an EA, including Categorical. (and NOT a Series)

cc @TomAugspurger

Reply:
Hrm I'll have to think about this... I'm hesitant to do this mainly because this code seems to rely on the fact that it always outputs an ndarray, which is why groupby doesn't work with sparse data.

The problem really is with classes that implement their own unique(). It usually outputs something that isn't ndarray.

Also I don't want to call unique on the class and cast it back to ndarray cause that feels sloppy. I'll think of something :) thanks!

Reply:
@hexgnu have you been following the ExtensionArray stuff? Eventually `SparseArray` should be an `ExtensionArray` like `Categorical`. Eventually things like `groupby`, `factorize`, etc will all work correctly on EAs.

> Also I don't want to call unique on the class and cast it back to ndarray cause that feels sloppy. 

SparseArray.unique() should return a SparseArray, just like Categorical.unique returns a Categorical.